### ISSUE TITLE
kubelet process restart leads to oscillation in Cloud LoadBalancer targets

### ISSUE THREAD

**Issue by zuzzas** [kubernetes/kubernetes#102367]
#### What happened:

If I were to restart kubelet on one or multiple nodes, it'd immediately set NotReady condition, and after checking PLEG healthiness, it will revert the Node object's condition to Ready. This leads to some thrashing in the list of targets for cloud LoadBalancers. Here's [the code](https://github.com/kubernetes/cloud-provider/blob/471c69a397d928c8fbed526379e600db8a16d51f/controllers/service/controller.go#L679-L684) that checks Node readiness in the cloud-provider.

This cannot be traced on some clouds, but on AWS (slow Initial health checks), it creates some issues on bulk kubelet restarts. This hits hard I were to do a rolling upgrade of kubelets (changing systems parameters or upgrading kubelet version) on multiple nodes.

#### What you expected to happen:

kubelet does not set a NotReady condition immediately after starting up or write some safety feature into the cloud-provider against oscillating NotReady conditions.

#### How to reproduce it (as minimally and precisely as possible):

`systemctl restart kubelet.service`

/sig node
/sig cloud-provider

#### Environment:
- Kubernetes version (use `kubectl version`): 

    ```Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
    Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
    ```
    
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`):

    ```
    NAME="Ubuntu"
    VERSION="18.04.3 LTS (Bionic Beaver)"
    ID=ubuntu
    ID_LIKE=debian
    PRETTY_NAME="Ubuntu 18.04.3 LTS"
    VERSION_ID="18.04"
    HOME_URL="https://www.ubuntu.com/"
    SUPPORT_URL="https://help.ubuntu.com/"
    BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
    PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
    VERSION_CODENAME=bionic
    UBUNTU_CODENAME=bionic
    ```

- Kernel (e.g. `uname -a`):

    ```
    Linux test-master-0 5.3.0-51-generic #44~18.04.2-Ubuntu SMP Thu Apr 23 14:27:18 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
    ```

- Install tools: kubeadm

**Comment by amshankaran** [comment]
@zuzzas  I'm also facing a similar issue. Could you please specify the time difference between the NodeNotReady state to NodeReady state? Have you enabled IPv6 in your cluster?

**Comment by zuzzas** [comment]
>Could you please specify the time difference between the NodeNotReady state to NodeReady state?
<details>
  <summary>literally one second</summary>
  
- apiVersion: v1
  count: 1
  eventTime: null
  firstTimestamp: "2021-05-28T09:20:48Z"
  involvedObject:
    kind: Node
    name: test-master-0
    uid: test-master-0
  kind: Event
  lastTimestamp: "2021-05-28T09:20:48Z"
  message: 'Node test-master-0 status is now: NodeNotReady'
  metadata:
    creationTimestamp: "2021-05-28T09:20:48Z"
    name: test-master-0.16833087454d654b
    namespace: default
    resourceVersion: "3262849"
    uid: d392c2ae-9a9e-490c-b999-c4e9eeba8aa4
  reason: NodeNotReady
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: kubelet
    host: test-master-0
  type: Normal
- apiVersion: v1
  count: 1
  eventTime: null
  firstTimestamp: "2021-05-28T09:20:49Z"
  involvedObject:
    kind: Node
    name: test-master-0
    uid: test-master-0
  kind: Event
  lastTimestamp: "2021-05-28T09:20:49Z"
  message: 'Node test-master-0 status is now: NodeReady'
  metadata:
    creationTimestamp: "2021-05-28T09:20:49Z"
    name: test-master-0.16833087858d0dcc
    namespace: default
    resourceVersion: "3262861"
    uid: 4db24e7b-f1e8-477f-a3e4-bcd8bda3375b
  reason: NodeReady
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: kubelet
    host: test-master-0
  type: Normal
</details>

>Have you enabled IPv6 in your cluster?
Nope.

**Comment by endocrimes** [comment]
/remove-sig node

from a node perspective this is somewhat expected behavior

**Comment by thockin** [comment]
I'm adding node back - while this is "expected" it's not "helpful" :)

It's not clear to me what the RIGHT answer is, but I do think it involve kubelet.

Does anyone have a link handy to where kubelet sets NotReady and why?

**Comment by matthyx** [comment]
This is a duplicate of #100277 

@thockin I am happy to discuss alternatives, but setting anything other than ready=false at kubelet start would have unexpected consequences...

/triage accepted
/assign

**Comment by Dingshujie** [comment]
> I'm adding node back - while this is "expected" it's not "helpful" :)
> 
> It's not clear to me what the RIGHT answer is, but I do think it involve kubelet.
> 
> Does anyone have a link handy to where kubelet sets NotReady and why?

@thockin, in some case, when kubelet start up, it call ReadyCondition to set Node Condition. Ready Condtion will check runtime, network, stoarge and so on.
when first time check runtime status, runtimeState.lastBaseRuntimeSync is not set, so may call runtimeErrors return container runtime status check may not have completed yet, this may lead to NotReady


```
func (s *runtimeState) runtimeErrors() error {
	s.RLock()
	defer s.RUnlock()
	errs := []error{}
	if s.lastBaseRuntimeSync.IsZero() {
		errs = append(errs, errors.New("container runtime status check may not have completed yet"))
	} else if !s.lastBaseRuntimeSync.Add(s.baseRuntimeSyncThreshold).After(time.Now()) {
		errs = append(errs, errors.New("container runtime is down"))
	}
	for _, hc := range s.healthChecks {
		if ok, err := hc.fn(); !ok {
			errs = append(errs, fmt.Errorf("%s is not healthy: %v", hc.name, err))
		}
	}
	if s.runtimeError != nil {
		errs = append(errs, s.runtimeError)
	}

	return utilerrors.NewAggregate(errs)
}
```
lastBaseRuntimeSync is set when initializeRuntimeDependentModules function call complete, and it will start cadvisor, container manager, eviction manager and so on. and cadvisor start cost 1 seconds when node have 50 or above pods.

**Comment by Dingshujie** [comment]
> This is a duplicate of #100277
> 

#100277 is about pod status change when  restart kubelet

**Comment by matthyx** [comment]
Do you mind taking a look at it?

/assign @Dingshujie

**Comment by Dingshujie** [comment]
> Do you mind taking a look at it?
> 
> /assign @Dingshujie
cadvisor detect all subcontainers when start. detect all containers will scan all cgroup path recursively.
if  node have 50 or above pods, it will cost a lot time, so kubelet will set node not ready at first

**Comment by thockin** [comment]
I understand why we set it, but I am saying it's not helpful.  What bad thing is being avoided by marking the node not ready?  If the workloads are still running (kubelet just restarted) then it doesn't matter that it costs a lot of time - the workloads are fine.

**Comment by zuzzas** [comment]
How can I help with pushing this issue forward? Should I go on Slack and pound my chest shouting justifications for my initial inquiry?

**Comment by thockin** [comment]
@matthyx said:

> This is a duplicate of https://github.com/kubernetes/kubernetes/issues/100277
> 
> @thockin I am happy to discuss alternatives, but setting anything other than ready=false at kubelet start would have unexpected consequences...

#100277 seems to be decided wrongly, to me.  I'm going to re-open that.

Can you explain to me what the unexpected consequences are?

@zuzzas In https://github.com/kubernetes/kubernetes/pull/109706 we are covering a different symptom but I think the "real" fix is related.

https://github.com/kubernetes/kubernetes/pull/109706#issuecomment-1179336082 in particular.

TL;DR: NodeReadiness should not factor into LBs (just as it does not factor into Endpoints)

**Comment by thockin** [comment]
Assigning to myself so I can track it better

**Comment by SergeyKanzhelev** [comment]
this should be fixed in a KEP: https://github.com/kubernetes/enhancements/pull/5493
